File: 36.有效的数独/36.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */

// eslint-disable-next-line no-unused-vars
const isValidSudoku = function(board) {
    let ch;

    for (let i = 0; i < 9; i++) {
        const rowNums = []
        const colNums = []
        const cubeNums = []

        for (let j = 0; j < 9; j++) {
            ch = board[i][j]
            if (ch !== '.') {
                if (rowNums.includes(ch)) return false
                rowNums.push(ch)
            }

            ch = board[j][i]
            if (ch !== '.') {
                if (colNums.includes(ch)) return false
                colNums.push(ch)
            }

            const row = Math.floor(i / 3) * 3 + Math.floor(j / 3)
            const col = (i % 3) * 3 + (j % 3)
            ch = board[row][col]
            if (ch !== '.') {
                if (cubeNums.includes(ch)) return false
                cubeNums.push(ch)
            }
        }
    }

    return true
}

File: algorithms/36. 有效的数独/36.py
from typing import List


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        for i in range(9):
            rowNums = []
            colNums = []
            cubeNums = []

            for j in range(9):
                ch = board[i][j]
                if (ch != '.'):
                    if ch in rowNums:
                        return False
                    rowNums.append(ch)

                ch = board[j][i]
                if (ch != '.'):
                    if ch in colNums:
                        return False
                    colNums.append(ch)

                row = i // 3 * 3 + j // 3
                col = i % 3 * 3 + j % 3
                ch = board[row][col]
                if (ch != '.'):
                    if (ch in cubeNums):
                        return False
                    cubeNums.append(ch)

        return True
